internal/sys/signals.go

```diff
@@ -0,0 +1,41 @@
+package sys
+
+import (
+	"fmt"
+	"runtime"
+
+	"github.com/cilium/ebpf/internal/unix"
+)
+
+var profSet unix.Sigset_t
+
+func init() {
+	if err := unix.SigsetAdd(&profSet, unix.SIGPROF); err != nil {
+		panic(fmt.Errorf("creating signal set: %w", err))
+	}
+}
+
+// MaskProfilerSignal locks the calling goroutine to its underlying OS thread
+// and adds SIGPROF to the thread's signal mask. This prevents pprof from
+// interrupting expensive syscalls like e.g. BPF_PROG_LOAD.
+//
+// Call defer sys.UnmaskProfilerSignal() to reverse the operation.
+func MaskProfilerSignal() {
+	runtime.LockOSThread()
+
+	if err := unix.PthreadSigmask(unix.SIG_BLOCK, &profSet, nil); err != nil {
+		panic(fmt.Errorf("masking profiler signal: %w", err))
```

Comment:
Maybe it doesn't matter, but with your proposed API panicking here will never execute `UnlockOSThread()`, because `UnmaskProfilerSignal` hasn't been deferred yet.

Easiest fix is a manual `UnlockOSThread` here I guess?

Reply:
True, that's the least we could do! An alternative could be `runtime.Goexit()`, as terminating a locked goroutine destroys its underlying thread. That wouldn't allow us to return error or panic, we'd have to log to stdout/err, so no go.